### ISSUE TITLE
'our sub' segfaults perl if the sub did not compile

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12823]
Migrated from [rt.perl.org#116981](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116981) (status was 'resolved')

Searchable as RT116981$

**Comment by p5pRT** [comment]
### From PeterCMartini@GMail.com

perl5\.17\.9 \-c \-Mfeature=lexical\_subs \-e 'our sub foo\{my 42\} foo'

\(Or blead\)

Crashes on a null pointer in&#8203;:

Program received signal SIGSEGV\, Segmentation fault\.
0x080be411 in Perl\_yylex \(my\_perl=0x82f5008\) at toke\.c&#8203;:7087
7087                        cv = lex ? GvCV\(gv\) : rv2cv\_op\_cv\(rv2cv\_op\, 0\);

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Summary of my perl5 (revision 5 version 17 subversion 9) configuration:
  Derived from: 250ef6a986d921a479be2c33b788d2aaf68438dd
  Platform:
    osname=linux, osvers=3.2.0-37-generic, archname=i686-linux-thread-multi
    uname='linux pmlinlaptop 3.2.0-37-generic #58-ubuntu smp thu jan
24 15:28:57 utc 2013 i686 i686 i386 gnulinux '
    config_args='-Dusedevel -DDEBUGGING -Dusethreads -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/i386-linux-gnu /lib/../lib
/usr/lib/i386-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib
-fstack-protector'


Characteristics of this binary (from libperl):
  Compile-time options: DEBUGGING HAS_TIMES MULTIPLICITY PERLIO_LAYERS
                        PERL_DONT_CREATE_GVSV PERL_IMPLICIT_CONTEXT
                        PERL_MALLOC_WRAP PERL_NEW_COPY_ON_WRITE
                        PERL_PRESERVE_IVUV PERL_TRACK_MEMPOOL PERL_USE_DEVEL
                        USE_ITHREADS USE_LARGE_FILES USE_LOCALE
                        USE_LOCALE_COLLATE USE_LOCALE_CTYPE
                        USE_LOCALE_NUMERIC USE_PERLIO USE_PERL_ATOF
                        USE_REENTRANT_API
  Locally applied patches:
        uncommitted-changes
  Built under linux
  Compiled at Feb 19 2013 01:30:54
  @INC:
    /usr/local/lib/perl5/site_perl/5.17.9/i686-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.17.9
    /usr/local/lib/perl5/5.17.9/i686-linux-thread-multi
    /usr/local/lib/perl5/5.17.9
    /usr/local/lib/perl5/site_perl
    .

```
</details>

**Comment by p5pRT** [comment]
### From PeterCMartini@GMail.com

Digging into this\, I also realize&#8203;:

perl \-Mfeature=lexical\_subs \-e 'my sub foo \{ foo\(1\);\} foo;'
Undefined subroutine &main&#8203;::foo called at \-e line 1\.

Is that intentional?

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Digging into this\, I also realize&#8203;:

perl \-Mfeature=lexical\_subs \-e 'my sub foo \{ foo\(1\);\} foo;'
Undefined subroutine &main&#8203;::foo called at \-e line 1\.

Is that intentional?

**Comment by p5pRT** [comment]
PeterCMartini@GMail.com - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @doy

On Wed\, Feb 27\, 2013 at 11&#8203;:05&#8203;:42PM \-0800\, Peter Martini via RT wrote&#8203;:
> Digging into this\, I also realize&#8203;:
> 
> perl \-Mfeature=lexical\_subs \-e 'my sub foo \{ foo\(1\);\} foo;'
> Undefined subroutine &main&#8203;::foo called at \-e line 1\.
> 
> Is that intentional?

It is consistent with how lexical variables work elsewhere \- new
lexicals are not available to be used until after the statement they are
declared in is finished&#8203;:

&nbsp;  $ perl \-E'use strict; say\(my $foo = 1\, $foo\)'
&nbsp;  Global symbol "$foo" requires explicit package name at \-e line 1\.
&nbsp;  Execution of \-e aborted due to compilation errors\.

On the other hand\, it's not consistent with how non\-lexical subs work\.
This is probably one of the cases where a decision had to be made about
what to be consistent with\, and I think in most cases\, the choice was to
be consistent with the equivalent long\-hand version\, so
'my $foo = sub \{ $foo\->\(1\) \}' in this case \(which also doesn't work\)\.
It's possibly useful to note that \_\_SUB\_\_ will work in both of these
cases\.

That said\, I do agree that it's confusing\, and an argument could
possibly be made for doing it the other way\, although I don't know what
the implications for the implementation would be\.

\-doy

**Comment by p5pRT** [comment]
### From PeterCMartini@GMail.com

On Thu\, Feb 28\, 2013 at 2&#8203;:15 AM\, Jesse Luehrs \<doy@&#8203;tozt\.net> wrote&#8203;:
> On Wed\, Feb 27\, 2013 at 11&#8203;:05&#8203;:42PM \-0800\, Peter Martini via RT wrote&#8203;:
>> Digging into this\, I also realize&#8203;:
>>
>> perl \-Mfeature=lexical\_subs \-e 'my sub foo \{ foo\(1\);\} foo;'
>> Undefined subroutine &main&#8203;::foo called at \-e line 1\.
>>
>> Is that intentional?
>
> It is consistent with how lexical variables work elsewhere \- new
> lexicals are not available to be used until after the statement they are
> declared in is finished&#8203;:
>
>   $ perl \-E'use strict; say\(my $foo = 1\, $foo\)'
>   Global symbol "$foo" requires explicit package name at \-e line 1\.
>   Execution of \-e aborted due to compilation errors\.
>
> On the other hand\, it's not consistent with how non\-lexical subs work\.
> This is probably one of the cases where a decision had to be made about
> what to be consistent with\, and I think in most cases\, the choice was to
> be consistent with the equivalent long\-hand version\, so
> 'my $foo = sub \{ $foo\->\(1\) \}' in this case \(which also doesn't work\)\.
> It's possibly useful to note that \_\_SUB\_\_ will work in both of these
> cases\.
>
> That said\, I do agree that it's confusing\, and an argument could
> possibly be made for doing it the other way\, although I don't know what
> the implications for the implementation would be\.
>
> \-doy

Ah\. point taken\, and you're precisely right\, I was thinking in terms
of subs instead of lexicals\. And it is documented\, so sorry for the
noise\.

**Comment by p5pRT** [comment]
### From PeterCMartini@GMail.com

Patch attached to make our sub behave a little better in the case of
syntax errors\.

If an entry is found in the pad\, but that gv is NULL\, simply pretend
that it wasn't found at all\.

Tests added; they behave the same way whether the subs are declared as
"sub" or "our sub"\.

Review appreciated\, since my modus operandi right now is mainly poke it
and see if I can make it break\.

"""
All tests successful\.
u=17\.70  s=4\.68  cu=1141\.10  cs=77\.78  scripts=2311  tests=618983
"""

**Comment by p5pRT** [comment]
### From PeterCMartini@GMail.com

<details><summary>0001-Fix-for-RT-116981.-If-a-sub-is-declared-with-our.patch</summary>

```diff
From 9be7b128444b7125ea04c39c69d63bee1019c6f6 Mon Sep 17 00:00:00 2001
From: Peter Martini <PeterCMartini@GMail.com>
Date: Thu, 28 Feb 2013 04:44:39 -0500
Subject: [PATCH] Fix for RT 116981. If a sub is declared with our, the name
 is added to the stash early, and left with a NULL ptr if
 there's a syntax error while compiling it.

Since the only time it becomes an issue is when that same name
is used in the same scope after a syntax error, what happens in
the pad is not particularly important.  The simple fix is to
simply fall back to treating it like a bareword, and pretending
it was never added to the pad in the first place.
---
 t/cmd/lexsub.t |   15 ++++++++++++++-
 toke.c         |    4 ++++
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/t/cmd/lexsub.t b/t/cmd/lexsub.t
index 86c7e26..46bab03 100644
--- a/t/cmd/lexsub.t
+++ b/t/cmd/lexsub.t
@@ -8,7 +8,7 @@ BEGIN {
     *bar::like = *like;
 }
 no warnings 'deprecated';
-plan 128;
+BEGIN{plan 133;}
 
 # -------------------- Errors with feature disabled -------------------- #
 
@@ -95,6 +95,19 @@ sub bar::c { 43 }
   my $y = if if if;
   is $y, 42, 'our subs from other packages override all keywords';
 }
+# Make sure errors don't pollute the stash (see RT 116981)
+{
+  eval "our sub ln99{!} ln99(1)";
+  eval "ln99(1)";
+  like $@, "Undefined subroutine &main::ln99 called", "Bad definitions do not pollute the stash";
+  isnt $::{ln99}, -1, "No placeholder was entered";
+  our sub ln103;
+  is $::{ln103}, -1, "Placeholder was entered";
+  eval "our sub ln103{!} ln103(1)";
+  eval "ln103(1)";
+  like $@, "Undefined subroutine &main::ln103 called", "Bad definitions do not pollute the stash";
+  isnt $::{ln103}, -1, "Placeholder was removed";
+}
 
 # -------------------- state -------------------- #
 
diff --git a/toke.c b/toke.c
index aace60b..aa6bca0 100644
--- a/toke.c
+++ b/toke.c
@@ -6900,6 +6900,10 @@ Perl_yylex(pTHX)
 		    gv = gv_fetchsv(sv, GV_NOADD_NOINIT | SvUTF8(sv),
 				    SVt_PVCV);
 		    off = 0;
+		    if (!gv) {
+			sv = NULL;
+			goto just_a_word;
+		    }
 		}
 		else {
 		    rv2cv_op = newOP(OP_PADANY, 0);
-- 
1.7.9.5


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Patch attached to make our sub behave a little better in the case of
syntax errors\.

If an entry is found in the pad\, but that gv is NULL\, simply pretend
that it wasn't found at all\.

Tests added; they behave the same way whether the subs are declared as
"sub" or "our sub"\.

Review appreciated\, since my modus operandi right now is mainly poke it
and see if I can make it break\.

"""
All tests successful\.
u=17\.70  s=4\.68  cu=1141\.10  cs=77\.78  scripts=2311  tests=618983
"""

**Comment by p5pRT** [comment]
### From PeterCMartini@GMail.com

Minor update to the previous patch\, it should obviously free the sv
before setting the pointer to it to null\.

**Comment by p5pRT** [comment]
### From PeterCMartini@GMail.com

<details><summary>0001-Fix-for-RT-116981.-If-a-sub-is-declared-with-our.patch</summary>

```diff
From 9be7b128444b7125ea04c39c69d63bee1019c6f6 Mon Sep 17 00:00:00 2001
From: Peter Martini <PeterCMartini@GMail.com>
Date: Thu, 28 Feb 2013 15:44:39 -0500
Subject: [PATCH 1/2] Fix for RT 116981. If a sub is declared with our, the
 name is added to the stash early, and left with a NULL
 ptr if there's a syntax error while compiling it.

Since the only time it becomes an issue is when that same name
is used in the same scope after a syntax error, what happens in
the pad is not particularly important.  The simple fix is to
simply fall back to treating it like a bareword, and pretending
it was never added to the pad in the first place.
---
 t/cmd/lexsub.t |   15 ++++++++++++++-
 toke.c         |    5 +++++
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/t/cmd/lexsub.t b/t/cmd/lexsub.t
index 86c7e26..46bab03 100644
--- a/t/cmd/lexsub.t
+++ b/t/cmd/lexsub.t
@@ -8,7 +8,7 @@ BEGIN {
     *bar::like = *like;
 }
 no warnings 'deprecated';
-plan 128;
+BEGIN{plan 133;}
 
 # -------------------- Errors with feature disabled -------------------- #
 
@@ -95,6 +95,19 @@ sub bar::c { 43 }
   my $y = if if if;
   is $y, 42, 'our subs from other packages override all keywords';
 }
+# Make sure errors don't pollute the stash (see RT 116981)
+{
+  eval "our sub ln99{!} ln99(1)";
+  eval "ln99(1)";
+  like $@, "Undefined subroutine &main::ln99 called", "Bad definitions do not pollute the stash";
+  isnt $::{ln99}, -1, "No placeholder was entered";
+  our sub ln103;
+  is $::{ln103}, -1, "Placeholder was entered";
+  eval "our sub ln103{!} ln103(1)";
+  eval "ln103(1)";
+  like $@, "Undefined subroutine &main::ln103 called", "Bad definitions do not pollute the stash";
+  isnt $::{ln103}, -1, "Placeholder was removed";
+}
 
 # -------------------- state -------------------- #
 
diff --git a/toke.c b/toke.c
index aace60b..aa6bca0 100644
--- a/toke.c
+++ b/toke.c
@@ -6900,6 +6900,11 @@ Perl_yylex(pTHX)
 		    gv = gv_fetchsv(sv, GV_NOADD_NOINIT | SvUTF8(sv),
 				    SVt_PVCV);
 		    off = 0;
+		    if (!gv) {
+			sv_free(sv);
+			sv = NULL;
+			goto just_a_word;
+		    }
 		}
 		else {
 		    rv2cv_op = newOP(OP_PADANY, 0);
-- 
1.7.9.5


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Minor update to the previous patch\, it should obviously free the sv
before setting the pointer to it to null\.

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Feb 28\, 2013 at 12&#8203;:53&#8203;:32PM \-0800\, Peter Martini via RT wrote&#8203;:
> Minor update to the previous patch\, it should obviously free the sv
> before setting the pointer to it to null\.

Looks plausible\. Thanks\, applied as
89e006ae4e39db68ad35c878eb6e6de83ebd8ec9\.

\-\- 
I before E\. Except when it isn't\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Mar 02 16&#8203;:26&#8203;:27 2013\, davem wrote&#8203;:
> On Thu\, Feb 28\, 2013 at 12&#8203;:53&#8203;:32PM \-0800\, Peter Martini via RT wrote&#8203;:
> > Minor update to the previous patch\, it should obviously free the sv
> > before setting the pointer to it to null\.
> 
> Looks plausible\. Thanks\, applied as
> 89e006ae4e39db68ad35c878eb6e6de83ebd8ec9\.

And now fixed in commit 3a74e0e282c for my and state subs\.


\-\- 

Father Chrysostomos
